src/charm.py

```diff
@@ -84,6 +84,10 @@ def _on_controller_relation_changed(self, event):
         self._stored.controllerData["identity-provider-url"] = event.relation.data[event.app].get("identity-provider-url", "")
         self._stored.controllerData["is-juju"] = event.relation.data[event.app]["is-juju"]
 
+        ip = str(self.model.get_binding(event.relation).network.ingress_address)
+        for relation in self.model.relations['controller']:
```

Comment:
Does this fail if `controller` is not defined....like if they only have an http relation but not a controller one?